### PR TITLE
Implement a better Array#product type

### PR DESCRIPTION
**Pull request by axyjo**
### Motivation
Array product currently returns a type that's technically correct, but not super useful.

See: [example](https://sorbet.run/#%23%20typed%3A%20strict%0A%0Afoo%20%3D%20T.let(%5B1%2C2%2C%203%5D%2C%20T%3A%3AArray%5BInteger%5D)%0Abar%20%3D%20T.let(%5B%22a%22%2C%20%22b%22%2C%22c%22%20%5D%2C%20T%3A%3AArray%5BString%5D)%0A%0AT.reveal_type(foo.product(bar))%20%23%20Expected%3A%20T%3A%3AArray%5B%5BInteger%2C%20String%5D%5D)

### Test plan

See included automated tests.

cc @akshay-stripe

### CODE REVIEW

File: core/types/calls.cc [hunk 1]
```diff
@@ -1916,6 +1916,45 @@ class Array_flatten : public IntrinsicMethod {
...
+        // Unwrap the array one time to get the element type (we'll rewrap it down at the bottom)
+        TypePtr element;
+        if (auto *ap = cast_type<AppliedType>(thisType)) {
+            ENFORCE(ap->klass == Symbols::Array() || ap->klass.data(ctx)->derivesFrom(ctx, Symbols::Array()));
+            ENFORCE(!ap->targs.empty());
+            element = ap->targs.front();
+        } else if (auto *tuple = cast_type<TupleType>(thisType)) {
+            element = tuple->elementType();
+        } else {
+            ENFORCE(false, "Array#product on unexpected type: {}", args.selfType->show(ctx));
```
Comment (DarkDimius) [inline core/types/calls.cc]: if this goes through, this should be a user-visible error.
Otherwise, this will be a user-visible segfault.

File: core/types/calls.cc [hunk 2]
```diff
@@ -1916,6 +1916,45 @@ class Array_flatten : public IntrinsicMethod {
...
+            auto argTyp = arg->type;
+            TypePtr el;
+            if (auto *ap = cast_type<AppliedType>(argTyp.get())) {
+                ENFORCE(ap->klass == Symbols::Array() || ap->klass.data(ctx)->derivesFrom(ctx, Symbols::Array()));
+                ENFORCE(!ap->targs.empty());
+                el = ap->targs.front();
+            } else if (auto *tuple = cast_type<TupleType>(argTyp.get())) {
+                el = tuple->elementType();
+            } else {
+                ENFORCE(false, "Array#product called with unexpected type: {}", el->show(ctx));
```
Comment (DarkDimius) [inline core/types/calls.cc]: ditto, but you can have _any_ type here in practice.

File: test/testdata/infer/product.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  [],
+  T::Array[T.untyped]
+)
+
+
+T.assert_type!(empty.product, T::Array[T.untyped])
+T.assert_type!(foo.product, T::Array[[String]])
+T.assert_type!(foo.product(mixed_type), T::Array[[String, T.any(String, Integer, Float)]])
+T.assert_type!(foo.product(bar), T::Array[[String, Integer]])
+T.assert_type!(foo.product(bar, foo), T::Array[[String, Integer, String]])
```
Comment (DarkDimius) [inline test/testdata/infer/product.rb]: we need negative tests too

**Comment by DarkDimius** [comment]
Could you give more motivation for this method "important enough" to be special cased?
Maintanence of Special-casing for every method substracts hours from Sorbet team building new features on ongoing basis(every year) and thus we need to make sure there's clear value in doing this over putting more time into other features.

Given that the previous behavior is correct(though imperfect) it's not entirely clear for me if this is worth ongoing maintenance. 
Said otherwise, has this caused incidents?

File: core/types/calls.cc (hunk 1 above)
Comment (jez) [reply core/types/calls.cc]: I actually think this one is fine to be an enforce, because this is on the `thisType` which since this intrinsic is on `Array#product`, we'll know that this `thisType` is always either an AppliedType of Array or a TupleType.

tl;dr, i think this ENFORCE is fine (especially because we have nearly the exact same check in `Array#flatten` and it has never fired)

File: core/types/calls.cc [hunk 4]
```diff
@@ -1916,6 +1916,45 @@ class Array_flatten : public IntrinsicMethod {
...
+            element = ap->targs.front();
+        } else if (auto *tuple = cast_type<TupleType>(thisType)) {
+            element = tuple->elementType();
+        } else {
+            ENFORCE(false, "Array#product on unexpected type: {}", args.selfType->show(ctx));
+        }
+
+        vector<TypePtr> unwrappedElems;
+        unwrappedElems.reserve(args.args.size() + 1);
+        unwrappedElems.emplace_back(element);
```
Comment (jez) [inline core/types/calls.cc]: Why even store `element` into an intermediary variable? It looks like we can just start with `unwrappedElems` at the top of the method, and `emplace_back` the unwrapped element as soon as we see it.

File: core/types/calls.cc (hunk 2 above)
Comment (jez) [reply core/types/calls.cc]: Take a look at `beginError` from the `Array#flatten` case to learn how it works. You probably want to either invent a new error code, or re-use the one that would normally result from mismatching an argument type:

https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20A%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20sig%20%7Bparams(x%3A%20Integer).void%7D%0A%20%20def%20self.foo(x)%3B%20end%0Aend%0A%0Adef%20main%0A%20%20A.foo(%22%22)%0Aend

^ "Expected Integer but found String" is error code 7002 which maps to `MethodArgumentMismatch`.

File: core/types/calls.cc [hunk 5]
```diff
@@ -1916,6 +1916,45 @@ class Array_flatten : public IntrinsicMethod {
...
+            } else if (auto *tuple = cast_type<TupleType>(argTyp.get())) {
+                el = tuple->elementType();
+            } else {
+                ENFORCE(false, "Array#product called with unexpected type: {}", el->show(ctx));
+            }
+
+            unwrappedElems.emplace_back(el);
+        }
+
+        res.returnType = Types::arrayOf(ctx, TupleType::build(ctx, unwrappedElems));
```
Comment (jez) [inline core/types/calls.cc]: I'm not really sure what `Array#product` does. Can you explain why this is the right type?

Comment (jez) [reply core/types/calls.cc]: Answering my own question: it looks like the number of args always corresponds to the length of the tuple, and other than that, the nesting depth of arrays in the result is always 2: the array of all combinations, containing tuples within.

File: test/testdata/infer/product.rb (hunk 3 above)
Comment (jez) [inline test/testdata/infer/product.rb]: There are no tests here that exercise the TupleType cases. These all operate on args and receivers that are Arrays.

I'd also like to see a test that calls product with like 3 or 4 args.

Comment (jez) [reply test/testdata/infer/product.rb]: And also a test that calls `product` with something that is not an array. You can expect type errors on certain lines in tests by putting `# error: (the message)` at the end of a line that should have an error.

**Comment by axyjo** [comment]
@DarkDimius Regarding the value: using `Array#product` at the moment is _very_ unergonomic if the arrays for which the cross product is being computed do not have the same type. This is a major quality of life improvement as it avoids having a `T.cast` for every element of the resulting array.

Furthermore, the current implementation also adds an unnecessary `T.nilable` around the result. For input arrays of lengths _n_ and _m_, `Array#product` will return _n*m_ elements, meaning that there won't be a `nil` in the result unless the arrays themselves have `nil` in them.

(Looking in to other PR feedback now)

**Comment by axyjo** [comment]
PTAL @jez @DarkDimius

File: core/types/calls.cc (hunk 2 above)
Comment (axyjo) [reply core/types/calls.cc]: Thanks, fixed!

File: test/testdata/infer/product.rb (hunk 3 above)
Comment (axyjo) [reply test/testdata/infer/product.rb]: Added multiple args, TupleType, as well as error cases

File: core/types/calls.cc (hunk 5 above)
Comment (axyjo) [reply core/types/calls.cc]: `Array#product` returns the cartesian products of the `this` array as well as all arrays passed in.

For arrays `A` and `B`, the result of `A.product(B)` is  `[(x, y) : x ∈ A, y ∈ B]`

**Comment by DarkDimius** [comment]
@jez, please take this. You have much more context that I do
